### PR TITLE
feat(gfx): white-preferred allocation for one white emitter (C3, #4041)

### PR DESCRIPTION
**Pull request by zackees**
The C++ side of the closed form established in #4187. Independent of the mapper PR (#4186) — this depends only on `device_solve.h`, which is on master.

With the white emitter at drive `w`, the RGB drives making up the difference are

```
d(w) = M⁻¹ · target − w · (M⁻¹ · white)
```

**Affine in `w`**, so each of the six bounds on the three RGB drives is one inequality in `w`, the feasible set is a single interval, and white-preferred is its upper end. `M⁻¹ · white` doesn't depend on the pixel and is precomputed at bind time, leaving one matrix multiply per pixel — no vertex enumeration, which is what A3/B11 rule out.

## Tested against the reference, not just for self-consistency

Vectors sampled from `ci/golden/color-reference-v1.json` for both `rgbw` and `non_d65_white`, spanning full-drive white, a saturated primary mix and the dark floor. The second device is the one that matters: nothing in the derivation assumed the white sat on the neutral axis, and that's what checks it.

Also asserted: the allocation actually *reproduces* its target when the four drives are pushed back through the emitter columns (what would break if the white column were subtracted with the wrong sign or scale), that the white level is genuinely maximal, and that out-of-gamut targets are refused rather than clamped — clamping is the mapper's job.

## Two bugs the tests caught, both mine

**The dark floor refused targets the device reproduces exactly.** Corpus target `(4, 1, 19)` solves to RGB drives `(0, −1, 1)`. That `−1` is pure rounding, but it dragged the interval's upper end below zero and the target came back as out of gamut. The upper end is now clamped up to zero, with the drive check deciding feasibility.

**Slack must not reach the bounds.** My first fix widened the interval bounds by the same 64 raw units the drive check tolerates. That looks symmetric and isn't: the policy *maximizes* white, so the allowance gets **spent** rather than tolerated — 64 units taken out of every RGB drive on every pixel. Measured as white = 886 where the reference says 0, on three separate vectors. The bounds are strict now; only the final drive check carries slack.

A third, smaller one was in the test itself: the maximality check nudged white by 256 units and then required a drive to escape by more than 256 — but a drive moves by `slope × nudge`, and blue carries only 0.072 of a unit of white here, so it moved 18. The nudge is now 1024 against a 2-unit window for division truncation.

## Cost, recorded rather than guessed at

Up to six s16.16 divisions per pixel — the only per-pixel division in the pipeline. Each bound can instead be compared by cross-multiplication, leaving one division for the chosen level, at the price of i64 multiplies that aren't obviously cheaper on an 8-bit target. Nobody has measured which wins, so the straightforward version ships and the header says so.

## Scope

Exactly one white emitter. Two are a different problem, and #4187 measured why: maximizing `w₁ + w₂` is a linear program over a polygon, and running this twice and keeping the better answer agrees with the reference on every corpus vector while being wrong on 85% of random reachable targets.

Part of #4034 / #4041.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01MLhWkMfzLjrnLTDMBE6Fj9

## Summary by CodeRabbit

- **New Features**
  - Added white-preferred RGBW color allocation for compatible emitter profiles.
  - Converts XYZ color targets into RGB and white drive levels while maximizing white output within valid drive limits.
  - Supports different white points, including D65 and D50.
  - Rejects unsupported targets and invalid emitter configurations instead of producing out-of-range results.

- **Bug Fixes**
  - Improved handling of very low white-emitter output to prevent invalid drive calculations.

- **Tests**
  - Expanded coverage for reference colors, boundary conditions, white-point variations, exact target reproduction, and invalid configurations.

### CODE REVIEW

File: src/fl/gfx/white_allocation.cpp.hpp [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+/// `(numerator << 16) / denominator`, in s16.16, with `denominator` non-zero.
+///
+/// Named for this file because .cpp.hpp files share a translation unit under
+/// the unity build, so an anonymous namespace does not isolate it.
+///
+/// The shift happens in i64: a numerator near i32's range shifted left by 16
+/// needs 47 bits, which is exactly the overflow this exists to avoid.
+i32 divideWhiteQ16(i32 numerator, i32 denominator) FL_NO_EXCEPT {
+    const i64 scaled = static_cast<i64>(numerator) << 16;
+    return static_cast<i32>(scaled / static_cast<i64>(denominator));
```
[the remark this answers is not in the document]
Comment (zackees) [reply src/fl/gfx/white_allocation.cpp.hpp]: Confirmed and fixed — same overflow class as the one caught in #4186, one module over. The quotient is now clamped before narrowing.

Clamped at full drive rather than at i32's edge, which is decision-preserving rather than a fudge: the caller only takes min against full drive and max against zero, so a bound outside +/-1.0 already means 'does not bind' or 'infeasible', and it still means that afterwards.

Regression vector added as you asked: a white emitter at a ten-thousandth of the primaries' luminance, which puts the slopes in the single raw units. It pins that its own fixture actually reaches that case (largest slope <= 16) so it cannot go quiet if the solve or the constants move. With the clamp removed, the allocation refuses that target outright and the test fails.

File: tests/fl/gfx/white_allocation.cpp [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        }
+        ++exercised;
+        i32 at_zero[3];
+        solveRgbDrivesQ16(allocation.rgb_solve, xyz, at_zero);
+        // The nudge has to be big enough that the *binding* drive clears the
+        // escape window, and the binding one may have a small slope: blue
+        // carries only 0.072 of a unit of white here, so a 256-unit nudge
+        // moves it 18 and would look like no escape at all. 1024 moves it 74,
+        // against a window of 2 units for the division's truncation.
+        const i32 nudged = drives[3] + 1024;
```
[the remark this answers is not in the document]
Comment (zackees) [reply tests/fl/gfx/white_allocation.cpp]: Right, and the arithmetic is exactly as you describe — with the smallest slope at 0.072 a 1024-unit nudge against a 2-unit window accepts about 1022 units of under-allocation, and the reconstruction test cannot see it because the RGB drives absorb the difference.

Replaced with the independently computed bound rather than an adaptive nudge: the test now derives the largest permissible white level in float from at_zero and per_white and compares the returned level against it, to 16 raw units.

Checked by mutation, and it demonstrates the gap you identified. Under-allocating by 100 units sits inside the reference comparison's 256-unit tolerance, so it is invisible there; it now fails the maximality test. Under the old nudge it passed everything.
